src/streaming/thread.pool.cpp

```diff
@@ -0,0 +1,96 @@
+#include "thread.pool.hh"
+
+zarr::ThreadPool::ThreadPool(unsigned int n_threads,
+                             std::function<void(const std::string&)> err)
+  : error_handler_{ err }
+  , is_accepting_jobs_{ true }
+{
+    n_threads = std::clamp(
+      n_threads, 1u, std::max(std::thread::hardware_concurrency(), 1u));
+
+    for (auto i = 0; i < n_threads; ++i) {
+        threads_.emplace_back([this] { thread_worker_(); });
+    }
+}
+
+zarr::ThreadPool::~ThreadPool() noexcept
+{
+    {
+        std::scoped_lock lock(jobs_mutex_);
+        while (!jobs_.empty()) {
+            jobs_.pop();
+        }
+    }
+
+    await_stop();
+}
+
+bool
+zarr::ThreadPool::push_to_job_queue(JobT&& job)
+{
+    std::unique_lock lock(jobs_mutex_);
```

Comment:
A similar question arises about the need to mix different locking mechanisms. I'm not an expert when it comes to the locking API, so there might be a compelling reason for this approach that I'm overlooking.